src/main/java/net/snowflake/ingest/utils/ParameterProvider.java

```diff
@@ -149,20 +160,74 @@ private void setParameterMap(Map<String, Object> parameterOverrides, Properties
 
     this.updateValue(
         MAX_CHUNK_SIZE_IN_BYTES, MAX_CHUNK_SIZE_IN_BYTES_DEFAULT, parameterOverrides, props);
+
+    this.updateValue(MAX_CLIENT_LAG, MAX_CLIENT_LAG_DEFAULT, parameterOverrides, props);
+    this.updateValue(
+        MAX_CLIENT_LAG_ENABLED, MAX_CLIENT_LAG_ENABLED_DEFAULT, parameterOverrides, props);
   }
 
-  /** @return Longest interval in milliseconds between buffer flushes */
+  /**
+   * @return Longest interval in milliseconds between buffer flushes
+   */
   public long getBufferFlushIntervalInMs() {
-    Object val =
-        this.parameterMap.getOrDefault(
-            BUFFER_FLUSH_INTERVAL_IN_MILLIS, BUFFER_FLUSH_INTERVAL_IN_MILLIS_DEFAULT);
-    if (val instanceof String) {
-      return Long.parseLong(val.toString());
+    if (getMaxClientLagEnabled()) {
+      if (cachedBufferFlushIntervalMs != -1L) {
+        return cachedBufferFlushIntervalMs;
+      }
+      long lag = getMaxClientLagMs();
+      if (cachedBufferFlushIntervalMs == -1L) {
+        cachedBufferFlushIntervalMs = lag;
+      }
+      return cachedBufferFlushIntervalMs;
+    } else {
+      Object val =
+          this.parameterMap.getOrDefault(
+              BUFFER_FLUSH_INTERVAL_IN_MILLIS, BUFFER_FLUSH_INTERVAL_IN_MILLIS_DEFAULT);
+      return (val instanceof String) ? Long.parseLong(val.toString()) : (long) val;
+    }
+  }
+
+  private long getMaxClientLagMs() {
+    Object val = this.parameterMap.getOrDefault(MAX_CLIENT_LAG, MAX_CLIENT_LAG_DEFAULT);
+    if (!(val instanceof String)) {
+      return 1000;
+    }
+    String maxLag = (String) val;
+    String[] lagParts = maxLag.split(" ");
+    if (lagParts.length != 2
+        || (lagParts[0] == null || "".equals(lagParts[0]))
+        || (lagParts[1] == null || "".equals(lagParts[1]))) {
+      return 1000;
+    }
+    long unit;
```

Comment:
nit: imo unit is the second part, this can be named value or lag

Reply:
Let's use lag.